### PR TITLE
Add support for using any()/none() in tag expressions

### LINKED ISSUE #1679: Run only "untagged" tests
I request the ability to run only "untagged" tests.

Most tests will run with default arguments and should not require `@Tag("fast")`. Other tests require special needs or take a long time so we tag them (i.e. `@Tag("slow")`). I would prefer not to require the majority of tests to be tagged at all.

Current workarounds include parsing the code for all declared instances of `@Tag("*")` or using the [Launcher#discover](https://github.com/junit-team/junit5/blob/a425b2c52b92befc148962f90d1358014cb38aec/junit-platform-launcher/src/main/java/org/junit/platform/launcher/Launcher.java#L82) method in the build phase, such that all found tags can be excluded.

## Deliverables

- Provide a keyword (i.e. "all") for the `excludeTags` property to exclude all tests that are tagged **OR** some solution providing the equivalent result.

### PR DESCRIPTION
**Pull request by mmerdes**
## Overview

This PR introduces support for two special tag expressions, `none()` and `any()`.

Resolved #1679.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

File: platform-tests/src/test/java/org/junit/platform/launcher/TagIntegrationTests.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+import static org.junit.platform.launcher.TagIntegrationTests.TaggedTestCase.unTaggedWasExecuted;
+import static org.junit.platform.launcher.core.LauncherDiscoveryRequestBuilder.request;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+import org.junit.platform.launcher.core.LauncherDiscoveryRequestBuilder;
+import org.junit.platform.launcher.core.LauncherFactory;
+
+class TagIntegrationTests {
```
Comment (marcphilipp) [inline platform-tests/src/test/java/org/junit/platform/launcher/TagIntegrationTests.java]: Didn't we have an existing class for such tests?

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/tagexpression/EnhancedParser.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * @since 1.6
+ */
+class EnhancedParser implements TagExpressionParser {
+
+	private final TagExpressionParser target = new ShuntingYardBasedParser();
+
+	@Override
+	public ParseResult parse(String infixTagExpression) {
+		if (infixTagExpression != null) {
+			if (infixTagExpression.equals("any()"))
```
Comment (marcphilipp) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/tagexpression/EnhancedParser.java]: I don't think we ever use `if` without braces.

File: documentation/src/docs/asciidoc/user-guide/running-tests.adoc [hunk 3]
```diff
@@ -734,6 +734,9 @@ those supplied via the configuration file.
 Tag expressions are boolean expressions with the operators `!`, `&` and `|`. In addition,
 `(` and `)` can be used to adjust for operator precedence.
 
+Two special expressions are supported, `any()` and `none()` which select all tests _with_
```
Comment (marcphilipp) [inline documentation/src/docs/asciidoc/user-guide/running-tests.adoc]:
```suggestion
Two special expressions are supported, `any()` and `none()`, which select all tests _with_
```

File: documentation/src/docs/asciidoc/user-guide/running-tests.adoc [hunk 4]
```diff
@@ -734,6 +734,10 @@ those supplied via the configuration file.
...
+any tags at all, and all tests _without_ any tags, respectively.
+These specials expressions may be combined with other expressions just like normal tags.
```
Comment (marcphilipp) [inline documentation/src/docs/asciidoc/user-guide/running-tests.adoc]:
```suggestion
These special expressions may be combined with other expressions just like normal tags.
```

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/tagexpression/Parser.java [hunk 5]
```diff
@@ -19,7 +19,7 @@
 
 	private final Tokenizer tokenizer = new Tokenizer();
 
-	ParseResult parse(String infixTagExpression) {
+	public ParseResult parse(String infixTagExpression) {
```
Comment (marcphilipp) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/tagexpression/Parser.java]:
```suggestion
	ParseResult parse(String infixTagExpression) {
```

**Comment by Richargh** [comment]
Thank you for this PR. None of our unit tests have tags, all of our slower tests have tags and the number of tags keeps increasing. This solves a big problem for us.

**Pull request merged**